include/oneapi/dpl/internal/dynamic_selection_impl/auto_tune_policy.h

```diff
@@ -0,0 +1,271 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) 2023 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_AUTO_TUNE_POLICY_H
+#define _ONEDPL_AUTO_TUNE_POLICY_H
+
+#include <stdexcept>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <utility>
+#include <chrono>
+#include <limits>
+#include <vector>
+#include <type_traits>
+#include <tuple>
+#include "oneapi/dpl/internal/dynamic_selection_traits.h"
+#if _DS_BACKEND_SYCL != 0
+    #include "oneapi/dpl/internal/dynamic_selection_impl/sycl_backend.h"
+#endif
+namespace oneapi {
+namespace dpl {
+namespace experimental {
+#if _DS_BACKEND_SYCL != 0
+  template <typename Backend=sycl_backend, typename... KeyArgs>
+#else
+  template <typename Backend, typename... KeyArgs>
+#endif
+  class auto_tune_policy {
+
+    static constexpr double never_resample = 0.0;
+    static constexpr int use_best_resource = -1;
+
+    using wrapped_resource_t = typename std::decay_t<Backend>::execution_resource_t;
+    using size_type = typename std::vector<typename Backend::resource_type>::size_type;
+
+    using timing_t = uint64_t;
+
+    struct resource_with_index_t {
+      wrapped_resource_t r_;
+      size_type index_;
+    };
+
+    struct time_data_t {
+      uint64_t num_timings_ = 0;
+      timing_t value_ = 0;
+    };
+
+    struct tuner_t {
+      std::mutex m_;
+
+      std::chrono::steady_clock::time_point t0_;
+
+      timing_t best_timing_ = std::numeric_limits<timing_t>::max();
+      resource_with_index_t best_resource_;
+
+      const size_type max_resource_to_profile_;
+      uint64_t next_resource_to_profile_ = 0; // as index in resources
+
+      using time_by_index_t = std::map<size_type, time_data_t>;
+      time_by_index_t time_by_index_;
+
+      double resample_time_ = 0.0;
+
+      tuner_t(resource_with_index_t br, size_type resources_size, double rt)
+        : t0_(std::chrono::steady_clock::now()),
+          best_resource_(br),
+          max_resource_to_profile_(resources_size),
+          resample_time_(rt) {}
+
+      size_type get_resource_to_profile() {
+        std::lock_guard<std::mutex> l(m_);
+        if (next_resource_to_profile_ < 2*max_resource_to_profile_) {
+          // do everything twice
+          return next_resource_to_profile_++ % max_resource_to_profile_;
+        } else if (resample_time_ == never_resample) {
+          return use_best_resource;
+        } else {
+          auto now = std::chrono::steady_clock::now();
+          auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(now-t0_).count();
+          if (ms < resample_time_) {
+            return use_best_resource;
+          } else {
+            t0_ = now;
+            return next_resource_to_profile_ = 0;
```

Comment:
You should probably return 0 but store 1 to `next_resource_to_profile_`

Reply:
done